server/json_client.go

```diff
@@ -0,0 +1,105 @@
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/mysterium/node/identity"
+	"net/http"
+	"net/url"
+)
+
+const (
+	mysteriumAgentName       = "goclient-v0.1"
+	authenticationHeaderName = "Authorization"
+	authenticationSchemaName = "Signature"
+)
+
+var noSignature = identity.SignatureBytes(nil)
+
+type HttpClient interface {
+	DoGet(path string, values url.Values) (*http.Response, error)
+	DoPost(path string, body interface{}) (*http.Response, error)
+	DoSignedPost(path string, body interface{}, signer identity.Signer) (*http.Response, error)
+}
+
+//HttpTransport interface with single method do is extracted from net/transport.Client structure
+type HttpTransport interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+func NewJsonClient(baseUrl string, transport HttpTransport) *jsonHttpClient {
+	return &jsonHttpClient{
+		baseUrl,
+		transport,
+	}
+}
+
+type jsonHttpClient struct {
+	baseApiUrl string
+	transport  HttpTransport
+}
+
+func (jhc *jsonHttpClient) DoGet(path string, values url.Values) (*http.Response, error) {
+	pathWithQuery := fmt.Sprintf("%v?%v", path, values.Encode())
+	return jhc.executeRequest(http.MethodGet, pathWithQuery, nil, noSignature)
+
+}
+
+func (jhc *jsonHttpClient) DoPost(path string, body interface{}) (*http.Response, error) {
+	return jhc.doSignedPayloadRequest(http.MethodPost, path, body, noOpSigner{})
+}
+
+func (jhc *jsonHttpClient) DoSignedPost(path string, body interface{}, signer identity.Signer) (*http.Response, error) {
+	return jhc.doSignedPayloadRequest(http.MethodPost, path, body, signer)
+}
+
+func (jhc *jsonHttpClient) doSignedPayloadRequest(method string, path string, body interface{}, signer identity.Signer) (*http.Response, error) {
+	payloadJson, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	signature, err := signer.Sign(payloadJson)
```

Comment:
we cannot have a signedHttpClient as per our discussion - in that case signer will become a state and new httpClient will be needed to construct for each signer

Reply:
Signing actually has much simpler flow:
```
request := client.createPOSTRequest("v1/node_register", requestDto{})
client.signRequest(&request, signer)

response, err := client.httpClient.Do(request)

err = parseResponseJson(response, &dto.ProposalsResponse{})

```

I think most complication comes, because we mixed *request creating* + *request transporting*:
- `doGetRequest()`
- `doPostRequest()`
versus:
- `createGetRequest()`
- `signRequest(request)`
- `doRequest(request)` wich already exist in `http.Client.Do()`

Reply:
In that case JsonClient becomes requestFactory?

Reply:
I did not get idea of `JsonClient` much. I am just advocating for function with single responsibility